src/clj/rems/db/events.clj

```diff
@@ -51,9 +51,43 @@
   [event]
   (let [old-event (fix-event-from-db (first (db/get-application-event {:id (:event/id event)})))
         _ (assert old-event)
-        event (-> old-event
-                  (merge event)
-                  (dissoc :event/id))]
-    (fix-event-from-db (db/update-application-event! {:id (:event/id old-event)
-                                                      :application (:application/id event)
-                                                      :eventdata (event->json event)}))))
+        event (merge old-event event)]
+    (db/update-application-event! {:id (:event/id old-event)
+                                   :application (:application/id event)
+                                   :eventdata (event->json (dissoc event :event/id))})
+    event))
+
+(defn replace-event!
+  "Replaces an event on top of an old one.
+
+  Differs from `add-event!` in that it replaces an old event.
+  Differs from `update-event!` in that the event id is a new one.
+
+  Returns the event as it went into the db."
+  [event]
+  (let [old-event (fix-event-from-db (first (db/get-application-event {:id (:event/id event)})))
+        _ (assert old-event)
+        event (merge old-event event)]
+    (fix-event-from-db (db/replace-application-event! {:id (:event/id old-event)
+                                                       :application (:application/id event)
+                                                       :eventdata (event->json (dissoc event :event/id))}))))
+
+(defn add-event-with-compaction!
+  "Add `event` to database.
+
+  Consecutive `:draft-saved` events of `application` are compacted into one by replacing
+  the last event of `application` instead of creating a new one.
+
+  Returns the event as it went into the db."
+  [application event]
+  (let [last-event (-> application
+                       :application/events
+                       last)]
+    (if (and (= :application.event/draft-saved
+                (:event/type event)
+                (:event/type last-event))
+             (= (:application/id event)
+                (:application/id last-event)))
```

Comment:
Now that I look like it, yes we could. I added the condition because it didn't seem to work otherwise, but with a closer inspection it mostly should have and it was probably something else. Technically there can be `n` events returned from a single command, and they could affect different applications (like "copy as new" does), but here, we are considering the save which never does it.

Reply:
I think I'll leave the check here as it is cheap. So it's kind of regression protection if we choose to create more complex commands or use the function from elsewhere.

I don't want to fetch the application of each event here with `get-application-internal` as that is not so cheap (it fetches the events and rebuilds the view).